src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ResettableValueTaskSource.cs

```diff
@@ -207,75 +244,31 @@ void IValueTaskSource.GetResult(short token)
                 if (state == State.Ready)
                 {
                     _valueTaskSource.Reset();
-                    if (_finalTaskSource.TrySignal(out Exception? exception))
+                    _state = State.None;
+
+                    // Propagate the _finalTaskSource result into _valueTaskSource if completed.
+                    if (_finalTaskSource.Task.IsCompleted)
                     {
                         _state = State.Completed;
-
-                        if (exception is not null)
+                        if (_finalTaskSource.Task.IsCompletedSuccessfully)
                         {
-                            _valueTaskSource.SetException(exception);
+                            _valueTaskSource.SetResult(true);
                         }
                         else
                         {
-                            _valueTaskSource.SetResult(true);
+                            // We know it's always going to be a single exception since we're the ones setting it.
+                            _valueTaskSource.SetException(_finalTaskSource.Task.Exception?.InnerException!);
```

Comment:
Why inner exception? 🤔 Above we were just setting  `_finalTaskSource.TrySetException(exception)` to whatever exception was passed to `ResettableValueTaskSource.TrySetException(exception, final: true)`

Reply:
`_finalTaskSource.Task.Exception` is `AggregateException`, so I'm pulling the original exception from it via `InnerException`.